lib/transport/selenium-webdriver/browserstack/appAutomate.js

```diff
@@ -15,7 +19,67 @@ class AppAutomate extends BrowserStack {
     return this.desiredCapabilities;
   }
 
-  createDriver({options}) {
+  async uploadAppToBrowserStack(options) {
+    const {appUploadPath, appUploadUrl} = options;
+
+    const multiPartFormData = {};
+    if (appUploadPath) {
+      multiPartFormData['file'] = {
+        filePath: path.resolve(appUploadPath)
+      };
+    } else if (appUploadUrl) {
+      multiPartFormData['url'] = {
+        data: appUploadUrl
+      };
+    }
+
+    if (options['appium:app'] && !options['appium:app'].startsWith('bs://')) {
+      multiPartFormData['custom_id'] = {
+        data: options['appium:app']
+      };
+    }
+
+    // eslint-disable-next-line no-console
+    console.log(`Uploading app to BrowserStack from '${appUploadPath || appUploadUrl}'...`);
+
+    try {
+      const response = await this.sendHttpRequest({
+        url: 'https://api-cloud.browserstack.com/app-automate/upload',
+        method: 'POST',
+        use_ssl: true,
+        auth: {
+          user: this.username,
+          pass: this.accessKey
+        },
+        multiPartFormData
+      });
+
+      if (response.error) {
+        Logger.error(response.error);
```

Comment:
We need a more descriptive error message here, like "Uploading the app the BrowSerstack has failed because of: ...". And we can also add some help text and a link for more info (if applicable), using the actionable error messages framework. Same goes for the error logging in the `catch()` block

Reply:
Done.

![image](https://user-images.githubusercontent.com/39924567/214013563-288a8435-bb49-4074-b831-3b90baec7dd3.png)

Reply:
the url in the screenshot contains `undefined`.

Reply:
Fixed in [0a4ba71](https://github.com/nightwatchjs/nightwatch/pull/3573/commits/0a4ba719f3e4c2695105069b51db842ce76deab6). We should only show the build URL if a session with BrowserStack was successfully created.